docs/RFCs/delivering-different-features.md

```diff
@@ -0,0 +1,78 @@
+# Building and delivering different features 
+
+## Context
+
+Our current build system is hard coded based on the profiling / telemetry requirements.
+With the growth of libdatadog, we need to adjust our build system.
+
+The current commands are :
+
+```
+./build_profiling-profiling.sh <package_directory>
+```
+
+OR
+
+```
+./build_telemetry-ffi.sh <package_directory>
+```
+
+Releases are then created through CI pipelines.
+
+The issue is that we have several features:
+- telemetry
+- exporting profiles
+- aggregating profiles
+- crash tracking
+- symbolization (ongoing).
+
+Not every profiler / tracer needs every feature, so we need it here way of Building and delivering exactly what we need.
+
+## Not a solution
+
+### Delivering different libraries
+
+By delivering features in several libraries, we expose ourselves to duplicated Rust runtimes (with possibly different versions and behaviours).
+
+## Proposed solution
+
+### Building everything together 
+
+Building a single static library with all features bundled. This means that downstream builds would be in charge of removing the parts that are not required.
+By linking against a static library, the compiler will remove the parts that are not needed.
+
+Cons
+- Longer build times
+- Cross team dependencies to deliver artifacts (flaky tests)
+- Shared libraries are no longer an option
+A shared library with everything in it will be too big. You need to link against a static library to remove the unused parts of the library.
+- Longer build times 
+Some profilers / tracers require compiling from source. Building against static libraries means longer build times.
+
+Pros
++ Simple build pipeline
++ Easier to experiment with a new feature (it is already availalbe locally)
+
+### Features
+
+Create new pipelines with a select amount of features.
+For languages that need shared libraries, we should make sure we are able to select the features we publish.
+
+The way this would work:
+
+#### Build
+We have as "shell" crate that pulls in all of the libdatadog features.
+We select what we want at build time.
+
+```
+cargo build --features symbolizer telemetry crash-tracking
+```
+
+#### Delivery
+
+CI pipelines are created with the required features.
+
+## Recommended
+
+I think we can do both proposed solutions. For the languages that require shared libraries, we can have the feature solution.
```

Comment:
Arguably, the "build everything together" option doesn't really preclude the shared library approach. We could still produce the one static library, and leave turning that into a shared library into a downstream per-library concern.

For instance, considering Ruby:
* The main libdatadog release artifact could be a static library
* This static library would be the new input to the Ruby packaging scripts, which would build a shared library from it, prior to publishing it to rubygems.org
* Alternatively -- we could re-examine our assumptions regarding using a shared library for Ruby being the best solution, and see if we can make a static library work and what would be the trade-offs in doing so

Reply:
I feel this is a slightly different solution that I could list.

Reply:
Also, this forces ruby to bundle a rust pipeline. Does this not mean solving the feature issue downstream ?

Reply:
I added this to the list of options

Reply:
> Also, this forces ruby to bundle a rust pipeline. Does this not mean solving the feature issue downstream ?

Not in the way I was thinking... Let me try to clarify:

Current state:
1. Libdatadog release builds a static and a dynamic library for profiling, which gets uploaded as a tarball to GitHub
2. The scripts in the `ruby/` folder in the libdatadog repository take the GitHub release, keeps only the dynamic library, adds a few Ruby helpers, packages and uploads it as the "libdatadog" gem to rubygems.org
3. Downstream dd-trace-rb consumes the "libdatadog" gem, using the shared library inside

I'm thinking this may be an option:
1. Libdatadog release builds only a static library with everything, which gets uploaded as a tarball to GitHub
2. We update the scripts in the `ruby/` folder to take as an input the static library and to repackage it as a dynamic library containing only the APIs that Ruby makes use of. That dynamic library gets uploaded as the "libdatadog" gem to rubygems.org
3. Same as above -- probably no changes needed

Reply:
To clarify, I'm not necessarily arguing for this approach, mostly stating that it may be an option if we want to focus on producing a single static library and then try to base everything on it, rather than having multiple paths to building from source.

Reply:
That is exactly what I intended to describe with my 3rd solution. Mission failed ;-) 
I'll add your description to illustrate